presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -605,55 +590,88 @@ public int hashCode()
     private static class DynamicFilterContext
     {
         private final Map<DynamicFilterId, Domain> dynamicFilterSummaries = new ConcurrentHashMap<>();
-        private final Supplier<List<StageDynamicFilters>> dynamicFilterSupplier;
         private final Set<DynamicFilterId> dynamicFilters;
         private final Map<DynamicFilterId, SettableFuture<?>> lazyDynamicFilters;
         private final Set<DynamicFilterId> replicatedDynamicFilters;
         private final AtomicBoolean completed = new AtomicBoolean();
+        private final Map<StageId, Set<DynamicFilterId>> stageDynamicFilters = new ConcurrentHashMap<>();
+        private final Map<StageId, Integer> stageNumberOfTasks = new ConcurrentHashMap<>();
+        // when map value for given filter id is empty it means that dynamic filter has already been collected
+        // and no partial task domains are required
+        private final Map<DynamicFilterId, Map<TaskId, Domain>> taskDynamicFilters = new ConcurrentHashMap<>();
```

Comment:
The currency of this class isn't really clear.  The fields use concurrent data structures, but the methods make coordinated changes between these data structures without synchronization.  Maybe the methods are safe but there aren't any comments on why it is a safe or warnings to be careful about changes.

Reply:
It is safe since:
1. We are always adding information (task DFs) to `DynamicFilterService`. Order of how these information arrive doesn't matter.
2. Only when all required task DFa are collected, a summarized DF is constructed and future is done. This can happen only once.
3. There is no "delete" in `DynamicFilterService`, which makes order of execution arbitrary.

Do you think there should be a comment like this here?

Reply:
Added comments